libqtile/group.py

```diff
@@ -30,11 +30,120 @@
 
 import xcffib
 import xcffib.xproto
+from xcffib.xproto import StackMode
 
 from libqtile import hook, utils, window
 from libqtile.command_object import CommandObject
 from libqtile.log_utils import logger
 
+# TODO: 1. I cannot move focus from a "fullscreen"-window to a "normal" window and back.
+#          NOTE: That might be a "feature" of the floating layout.
+#       2. A fullscreen window can spawn its child-popups behind itself.
+#          NOTE: According to @ramnes we aren't managing popups yet; plus, same behaviour happens in the master branch.
+#       3. After restarting qtile, only the topmost window in a max layout is visible.
+#          NOTE: There are more weird things caused by restarts, and according to tych0 the root problem is that
+#                restarts should only be needed during development/testing.
+
+
+class _History(list):
```

Comment:
I am working on removing this and using `_ClientList` (in layout/base.py) instead. It does feel hacky, though, since that class approaches the "management" of clients very differently and the concepts clash a little.
Do you generally think that the clients should be managed through a bare list, or just that this class shouldn't be added, if possible?

Reply:
Yeah, I was thinking of reusing `ClientList` as well but then read the code again and it's indeed very hacky as it doesn't serve the same purpose. And I think this will stay hacky as long as we'll let the layouts handle focus and other similar stuff themselves. We could merge these two logics by extracting a lot of functionalities from the layouts, as I already proposed here: https://github.com/qtile/qtile/pull/1500#issuecomment-629820596. But this is definitely outside the scope of this PR so I'd rather just manage the clients through a simple list meanwhile.

Reply:
There might be more code with the class included, but the related code inside `Group` is easy to read as a result of it. I could move the relevant functions into `Group`, but then I can't reuse any of the code for the fake screens later. So I am a little torn on this.

Thanks for sharing your opinion on `_ClientList`. It looks like, if there weren't any layouts using it yet, it would be a good idea to merge the classes or at least have a common base. Either way, I'll drop trying to unite the two for now.

Reply:
Well, if you really think that a class is mandatory for this and for the future improvements you have in mind, we have to work with `_ClientList`. While it serves a different purpose (handling the clients in a layout and the layout's focus), it is still a list of clients, and it would be a shame to reimplement half the methods on both sides. What do you think of extracting what can be useful for us here into another class, and make `_ClientList` inherit from this newer class? We could have `ClientList` and `LayoutClientList` or `layouts.ClientList`, for example. It's not optimal but still better than two different classes that have half the methods in common, I think. I wish this was in another PR though.

Reply:
I'll try to remove the `_History`-class then, but I'll leave at least two of the methods around - likely in `Group`. One change this will lead to overall, though, is that `Group.windows` won't be a set any longer (I have to rely on the ordering) but a proper list. That does open up the possibility of accidentally "misusing" it in the future, since there won't be any protection against duplicate entries. Do you think that's acceptable?

I agree that an extracted base class is the most attractive option here, and that it should be handled in a different PR.

Reply:
The class is removed. Let me know what you think.